src/main/java/picard/arrays/illumina/IlluminaBPMFile.java

```diff
@@ -0,0 +1,181 @@
+package picard.arrays.illumina;
+
+import picard.PicardException;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * A class to parse the contents of an Illumina Bead Pool Manifest (BPM) file
+ *
+ * A BPM file contains metadata (including the alleles, mapping and normalization information) on an Illumina Genotyping Array
+ * Each type of genotyping array has a specific BPM .
+ *
+ */
+public class IlluminaBPMFile extends InfiniumDataFile {
+    private static final String BPM_IDENTIFIER = "BPM";
+
+    private String manifestName = null;
+    private String controlConfig = null;
+    private int numLoci = 0;
+    private IlluminaBPMLocusEntry[] locusEntries = null;
+
+    private int[] allNormalizationIds = null;
+    private Integer[] uniqueNormalizationIds = null;
+
+    public IlluminaBPMFile(DataInputStream stream) throws IOException {
+        super(stream, true);
+        parse();
+    }
+
+    /**
+     * Main parsing method.
+     *
+     * @throws IOException thrown when there is a problem reading the Bead Pool Manifest
+     */
+    private void parse() throws IOException {
+
+        stream.mark(0);
+
+        try {
+            final byte[] formatIdentifier = new byte[BPM_IDENTIFIER.length()];
+            for (int i = 0; i < formatIdentifier.length; i++) {
+                formatIdentifier[i] = parseByte();
+            }
+
+            final String identifier = new String(formatIdentifier);
+            setIdentifier(identifier);
+            if (!identifier.equals(BPM_IDENTIFIER)) {
+                throw new PicardException("Invalid identifier '" + identifier + "' for BPM file");
+            }
+            setFileVersion(parseByte());
+            if (getFileVersion() != 1) {
+                throw new PicardException("Unknown BPM version (" + getFileVersion() + ")");
+            }
+            int version = parseInt();
+            final int versionFlag = 0x1000;
+            if ((version & versionFlag) == versionFlag) {
+                version ^= versionFlag;
+            }
+            if (version > 5 || version < 3) {
+                throw new PicardException("Unsupported BPM version (" + version + ")");
+            }
+            manifestName = parseString();
+            if (version > 1) {
+                controlConfig = parseString();
+            }
+            numLoci = parseInt();
+
+            readData();
+        } finally {
+            stream.close();
+        }
+    }
+
+    private void readData() throws IOException {
+        // Skip the index block
+        stream.skipBytes(4 * numLoci);
+
+        // Read the names
+        String[] names = new String[numLoci];
+        for (int i = 0; i < numLoci; i++) {
+            names[i] = parseString();
+        }
+
+        // Read the normalization ids.
+        allNormalizationIds = parseByteArrayAsInts(numLoci);
+        Set<Integer> uniqueNormalizationIdsSet = new TreeSet<>();
```

Comment:
Could probably use a hashset here
```suggestion
        final Set<Integer> uniqueNormalizationIdsSet = new HashSet<>();
```